### ISSUE TITLE
slaves running out of memory

### ISSUE THREAD

**Issue by tfoote** [ros-infrastructure/ros_buildfarm#45]
Overnight we had a lot of failures due to one of the slaves running out of memory: 

Between 4 and 5am there are a lot of instances of: 
```
04:38:54 Sending build context to Docker daemon 
04:38:54 time="2015-03-18T11:38:55Z" level="fatal" msg="Error: Untar fork/exec /usr/bin/docker: cannot allocate memory" 
```

On 1 specific slave. 

Inspecting the slave now(while idle) shows very high memory: 
```
  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND                                         
  935 root      20   0 9220968 1.495g   7936 S   0.0 10.2   3312:38 docker                                          
 1339 jenkins+  20   0 8651680 402752   6580 S   0.0  2.6  45:09.64 java                                            
11442 proxy     20   0  657284 370000   2176 S   0.0  2.4  19:32.68 squid3  
```
This is an operational state. It must have been worse earlier.

On one of the other slaves which didn't have a problem the memory usage is noteably lower
```
  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND                                         
  928 root      20   0 6991428 1.093g   7932 S   0.0  7.4   4272:25 docker                                          
 1280 jenkins+  20   0 9077376 430072   5836 S   0.0  2.8  83:35.66 java                                            
26268 proxy     20   0  658616 369348   2144 S   0.0  2.4  20:23.54 squid3  
```

There is a closed ticket with exactly this error: https://github.com/docker/docker/issues/8539 which has been marked as a duplicate of: https://github.com/docker/docker/issues/9139

I'm not entirely convinced this is just the proposed buffer overflow marked as the fix. 

I think based on the times of the failures that during the problem we were also running the cleanup script. 

After restarting docker it's taking a lot less memory: 
```
  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND                                         
 1280 jenkins+  20   0 9078404 430004   5840 S   0.0  2.8  83:39.15 java                                            
 8335 proxy     20   0  381208  93248   5564 S   1.3  0.6   0:00.53 squid3                                          
 7630 root      20   0  577164  74272   8048 S   0.0  0.5   0:17.32 docker  
```

**Comment by tfoote** [comment]
@wjwwood FYI

**Comment by tfoote** [comment]
Saw this again today. 

```
  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND                                                                   
  935 root      20   0 9222248 6.281g   7980 S 100.1 42.8   5943:31 /usr/bin/docker -d -H unix:///var/run/docker.sock                         
 1339 jenkins+  20   0  9.869g 460780   5600 S   0.0  3.0  93:07.07 /usr/bin/java -jar /home/jenkins-slave/swarm-client-1.17-jar-with-depend+
```

Docker is running no containers at the moment.  And I cannot run any container: 
```
root@ip-172-31-3-61:~# docker run -ti ubuntu:trusty bash
FATA[0059] Error response from daemon: Cannot start container b7045ad0ac37799f73337ea70eefb0bc67da76873f7f34a7e27ca60d8370e58c: fork/exec /var/lib/docker/init/dockerinit-1.5.0: cannot allocate memory 
```
Memory is low but not exhausted. With a good amount in cached. 
```
root@ip-172-31-3-61:~# free -h
             total       used       free     shared    buffers     cached
Mem:           14G        14G       547M       640K       915M       4.4G
-/+ buffers/cache:       8.9G       5.8G
Swap:           0B         0B         0B
```

**Comment by tfoote** [comment]
Looking at this it appears with docker's large memory footprint, exec is refusing to fork because it would require 6.2Gb of space which was not available.

**Comment by tfoote** [comment]
I upgraded to test docker 1.6 rc2 and it's still leaking memory. 

```
  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND                                                                   
 4374 root      20   0 9091484 6.758g   8412 S 100.1 46.0 538:23.07 docker   
```

```
# docker --version
Docker version 1.6.0-rc2, build c5ee149
```

```
# docker info
Containers: 603
Images: 8297
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 9587
Execution Driver: native-0.2
Kernel Version: 3.13.0-44-generic
Operating System: Ubuntu 14.04.1 LTS
CPUs: 4
Total Memory: 14.69 GiB
Name: ip-172-31-4-7
ID: 6KW6:XMAA:BLMZ:F5U2:QMGY:JNZR:FCZ5:FVDU:GJUS:KSHQ:FSR3:AO62
WARNING: No swap limit support
```

**Comment by tfoote** [comment]
I have given feedback on 1.6 rc2 not fixing this problem at: https://github.com/docker/docker/issues/9139#issuecomment-87033577

**Comment by motin** [comment]
It this related to https://github.com/docker/docker/issues/12848 ? If so it is fixed in upcoming 1.7.0

**Comment by tfoote** [comment]
It does seem to be resolved now with docker 1.7+
